server/raft.go

```diff
@@ -3247,6 +3255,12 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 		n.updateLeadChange(false)
 	}
 
+	// Check if this a new heartbeat with a newer term but same index.
+	// If so we take on the new term and proceed.
+	if isNew && ae.pterm > n.pterm && ae.pindex == n.pindex && len(ae.entries) == 0 {
+		n.term, n.pterm = ae.pterm, ae.term
```

Comment:
I'm not sure that updating our own `pterm` in this way is safe. Shouldn't it represent what we believe the term was at our last local log entry?

Reply:
Yes however we aggressively send HBs and they can come before peer state in my testing over weekend.

Reply:
So without this the heartbeat causes us to get rid of last entry, enter catchuo, get it back exactly the same, then the peerstate from the takeover which moves pterm, etc..

Reply:
I think resetting the vote `term` is safe, but for `pterm` it seems like notching that up from the heartbeats means we lose our ability to tell if the leader's log is consistent with ours, i.e. if you have one node that has index 10 as belonging to term 5 and another node that has index 10 belonging to index 6, those are divergent logs and we'd rely on `pterm` to highlight that I think.

Reply:
I can remove it, the system will work just in the case of leader takeover from restart the path will be a bit slower with a catchup loop that is a no-op.

Reply:
Will remove for now and track down the HB trigger condition later.